### PR TITLE
Optimize stake accounts store

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem
Storing stake accounts suffers from RWLock contention on stake_cache.  That's why storing stake account (running highly parallel) takes much longer time than storing vote account (running single threaded).

#### Summary of Changes
Separate stake account store from reward redeem. Keep reward redeem running in parallel, batch the staking accounts to store and store the stake accounts in a single thread. 

**Result**
Store time reduced from 4.4s to 1.7s.

```
update_rewards_with_thread_pool_us=7230692i load_vote_and_stake_accounts_us=4454714i store_stake_accounts_us=1767486i
```
```
update_rewards_with_thread_pool_us=12077002i load_vote_and_stake_accounts_us=7272740i store_stake_accounts_us=4467960i
```
Fixes #

### CODE REVIEW

File: runtime/src/stakes.rs
Comment (HaoranYi) [reply runtime/src/stakes.rs:114]: didn't measure. But it reduce the number of lock and unlock instructions at least.

Comment (behzadnouri) [reply runtime/src/stakes.rs:114]: Let's remove it for now if there is no evidence that this has any significant impact.
You can benchmark `RwLock::write().unwrap()` in a `for` loop or something.

File: runtime/src/bank.rs
Comment (behzadnouri) [inline runtime/src/bank.rs:178]: I think you can move this struct inside the `pay_validator_rewards_with_thread_pool` function; It is not used anywhere else.

`pubkey` is ambitious. Please rename to `stake_pubkey`.
Same for `account -> stake_account`.

`balance -> stakers_reward`

File: runtime/src/bank.rs
Comment (behzadnouri) [inline runtime/src/bank.rs:3243]: Separate `filter`, `map`s are more readable, shorter and avoid `if` branches:
```rust
.filter(|r| r.balance > 0)
.map(|r| (r.pubkey, r.reward))
```

File: runtime/src/bank.rs
Comment (behzadnouri) [inline runtime/src/bank.rs:3198]: Can we name this the same as struct filed `post_balance`?

File: runtime/src/bank.rs
Comment (behzadnouri) [inline runtime/src/bank.rs:3185]: Can you please keep this comment either here or where it is more relevant now?

File: runtime/src/bank.rs
Comment (behzadnouri) [inline runtime/src/bank.rs:184]: does not have to be atomic.
Probably also the other ones, but those can be fixed in a separate commit.

File: runtime/src/bank.rs
Comment (behzadnouri) [inline runtime/src/bank.rs:3236]: `.into_iter()`

File: runtime/src/stakes.rs
Comment (HaoranYi) [reply runtime/src/stakes.rs:114]:
```
running 2 tests
test bench_stake_cache_write       ... bench:   8,758,355 ns/iter (+/- 3,730,077)
test bench_stake_cache_write_batch ... bench:   6,020,470 ns/iter (+/- 977,997)
```

Comment (HaoranYi) [reply runtime/src/stakes.rs:114]: benchmarks shows about 25% speed up with batch store.

Comment (HaoranYi) [reply runtime/src/stakes.rs:114]: https://github.com/HaoranYi/solana/commit/feaa161a9d7a4a66de49b6c146609440a824c798

Comment (behzadnouri) [reply runtime/src/stakes.rs:114]: Ok, that is like 3ms for 3,000 stake accounts. So 1us for each account;
Can you please try the benchmark with 600k stake account?

Comment (HaoranYi) [reply runtime/src/stakes.rs:114]: yeah. 
```
running 1 test
    write loop time: 26852ms
    write batch time: 26747ms
    test stakes::tests::test_timing_stake_cache_write ... ok
```
With 600K accounts, rust bench mark never converge. So I wrote a separate test to explicit time the exaction. They show similar timing. batch is slightly faster.

File: runtime/src/bank.rs
Comment (behzadnouri) [reply runtime/src/bank.rs:176]: any reason to keep this atomic?

File: runtime/src/stakes.rs
Comment (behzadnouri) [reply runtime/src/stakes.rs:114]: Ok, then I am leaning not to add this extra api point,
until there is evidence that it has significant impact for rewards calculation at epoch boundary.
(mainnet stake accounts are like ~520k).

Comment (HaoranYi) [reply runtime/src/stakes.rs]: OK. Removed.

**Review by jeffwashington — APPROVED** [review]
lgtm
I would hold off submitting until @behzadnouri or @jstarry approve.

File: runtime/src/bank.rs
Comment (behzadnouri) [inline runtime/src/bank.rs:3148]: Isn't `StakeReward.stake_reward` redundant?
Looks like we can just use `StakeReward.stake_reward_info.lamports`.

File: runtime/src/bank.rs
Comment (behzadnouri) [inline runtime/src/bank.rs:3144]: I see that this is an old code, but I think it is better to abort here rather than have any chance of integer overflow.
Can you please change this to `i64::try_from(stakers_reward).unwrap()`?

**Review by behzadnouri — APPROVED** [review]
lgtm

**Pull request merged**
